rfcs/004 - Continuous Two-Way Synchronization RFC.md

```diff
@@ -0,0 +1,163 @@
+```
+
+Similar to the enforce and track functions, it is essentially executing (on another postgres thread/process) a particular select statement in a loop with a sleep thrown in until success or it errors out if the condition is never met. That select statement would be something like:
+
+```sql
+SELECT count(*) > 0 FROM `table_name` WHERE id = `5` AND `cloud_id IS NOT NULL`
+```
+
+The parts in backticks are from the user-provided values. This is SQL injection, but it's their own database and we already allow arbitrary SQL statements to run on the database, so it doesn't affect the security situation in the slightest, even if it does immediately cause heart palpitations while looking at it. ;) This is also not blocking for the transition, but simply a pair of convenience functions that we should be able to provide relatively easily.
+
+With this, we now have enough to unblock the `code-` services, but we can continue from here to emulate the two One-Way automatic modes, we can take a page from Postgres' [Table and Row-Level Locking](https://www.postgresql.org/docs/15/explicit-locking.html).
+
+Naming these modes to be clearly understood will be important so "Anti-Chaos-Monkey" can't be what we go with here. ;) The current proposal for the name for the "Write-Only Automatic" style functions is "Enforce" as the user is asking IaSQL to enforce these particular rows or tables to always match the database, and the "Read-Only Automatic" style becomes "Track" as the user is asking IaSQL to track the upstream cloud. These various configuration choices have differing levels of scope and collisions between the rules can be confusing. How these are interpreted needs to be well-defined and queryable. So we need a new table to represent these states in the database, and then we should follow a few rules:
+
+1. More specific rules override less specific rules, so if we "track" cloud changes for a table, but "enforce" one particular row from that table, the "enforced" row will be "write-only automatic" while the rest of the records in the table will be "read-only automatic".
+2. Equal level contradictory rules should not exist, but if they can, we should err on the side of not breaking things, so "Read-Only Automatic" > "Two-Way Automatic" > "Write-Only Automatic" when dealing with a tie.
+
+That last one *shouldn't* be a problem for us, though, because we can define a new `iasql_mode` table like this:
+
+```
++-----------------------------------------+
+| iasql_mode                              |
++-----------------------------------------+
+| table_name NULLABLE VARCHAR             |
+| record_id  NULLABLE INT                 |
+| mode       ENUM(NORMAL, ENFORCE, TRACK) |
++-----------------------------------------+
+| UNIQUE (table_name, record_id)          |
+| FK (table_name => iasql_tables.table)   |
++-----------------------------------------+
```

Comment:
can we no longer change the behavior of the cron / commit with a `iasql_enforce_all` and `iasql_track_all` functions?

Reply:
Sure we can, just do:

```sql
INSERT INTO iasql_mode (mode) VALUES ('ENFORCE');
```

If you have `NULL` for the `table_name` and the `record_id` it would be interpreted as a wildcard "select all" in that category type of thing. I mention this in the paragraph below the schema table, but let me know if it's confusing and why?

Reply:
I added a bit more explanatory paragraphs here. I hope that helps?